### PR TITLE
Add middleware analysis tools, sample.

### PR DESCRIPTION
**Pull request by Tratcher**
Hosting#318 @avanderhoorn @davidfowl @lodejard 
Depends on:
https://github.com/aspnet/Hosting/commit/f8d2b9bcfb0ea3b77631c72526c13b4e774ec7c8
https://github.com/aspnet/HttpAbstractions/commit/0e03acfc5e54d81e7a2616882e8ef97299ed404a

Sample output:
```
info: Microsoft.AspNet.Hosting.Internal.HostingEngine[1]
      Request starting HTTP/1.1 GET http://localhost:5000/map/foo
MiddlewareStarting: RequestServicesContainerMiddleware; /map/foo
MiddlewareStarting: DeveloperExceptionPageMiddleware; /map/foo
MiddlewareStarting: IISPlatformHandlerMiddleware; /map/foo
MiddlewareStarting: UseInlineMiddleware; /map/foo
MiddlewareStarting: MapMiddleware; /map/foo
MiddlewareStarting: RunInlineMiddleware; /foo
MiddlewareFinished: RunInlineMiddleware; 200
MiddlewareFinished: MapMiddleware; 200
MiddlewareFinished: UseInlineMiddleware; 200
MiddlewareFinished: IISPlatformHandlerMiddleware; 200
MiddlewareFinished: DeveloperExceptionPageMiddleware; 200
MiddlewareFinished: RequestServicesContainerMiddleware; 200
info: Microsoft.AspNet.Hosting.Internal.HostingEngine[2]
      Request finished in 0.0047ms 200
verb: Microsoft.Net.Http.Server.WebListener[0]
      Sending Response: HTTP/1.1 200 OK; Headers: { Content-Length: 11; }
```

### CODE REVIEW

**Comment by Tratcher** [comment]
*ping* @davidfowl @avanderhoorn 
/cc: @muratg

**Comment by davidfowl** [comment]
NameNextMiddleware is bizzare doesn't feel right. Don't like it

**Comment by davidfowl** [comment]
So everything else is good except for `AnalysisApplicationBuilderExtensions.GetNextMiddlewareName(this)` and the whole concept of naming anonymous middleware. Just use the name of the delegate (like reflector).

**Comment by Tratcher** [comment]
NameNextMiddleware isn't just for anonymous middleware, It also lets you assign more meaningful names to any other middleware in your app. E.g. like when Identity has three different cookie auth middleware.

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            _next = next;
+            _diagnostics = diagnosticSource;
+            _middlewareName = middlewareName;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            if (_diagnostics.IsEnabled("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareStarting"))
+            {
+                _diagnostics.Write("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareStarting", new { name = _middlewareName, httpContext = httpContext, tickCount = Environment.TickCount });
```
Comment (lodejard) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs]: Don't put TickCount as event data - assume that the listeners will be using their own clocking system

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            }
+
+            try
```
Comment (lodejard) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs]: change to 
`
//write("starting")
try
{
  await _next(httpContext);
}
catch(exception)
{
  //write("exception")
  throw;
}
finally
{
  //write("finished")
}
`

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                if (_diagnostics.IsEnabled("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareFinished"))
+                {
+                    _diagnostics.Write("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareFinished", new { name = _middlewareName, httpContext = httpContext, tickCount = Environment.TickCount });
+                }
+            }
+            catch (Exception ex)
+            {
+                if (_diagnostics.IsEnabled("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareException"))
+                {
+                    _diagnostics.Write("Microsoft.AspNet.MiddlewareAnalysis.MiddlewareException", new { name = _middlewareName, exception = ex, tickCount = Environment.TickCount });
```
Comment (lodejard) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs]: shouldn't the exception get the httpcontext also?

File: samples/MiddlewareAnalysisSample/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            services.AddMiddlewareAnalysis();
+        }
+
+        public void Configure(IApplicationBuilder app, ILoggerFactory factory)
+        {
+            // Displays all log levels
+            factory.AddConsole(LogLevel.Verbose);
+
+            // Listen for middleware events and log them to the console.
+            var diagnosticListener = app.ApplicationServices.GetRequiredService<DiagnosticListener>();
```
Comment (lodejard) [inline samples/MiddlewareAnalysisSample/Startup.cs]: Take `DiagnosticListener` as an additional Configure parameter?

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            return new AnalysisBuilder(InnerBuilder.New());
+        }
+
+        public IApplicationBuilder Use(Func<RequestDelegate, RequestDelegate> middleware)
+        {
+            string middlewareName = AnalysisApplicationBuilderExtensions.GetNextMiddlewareName(this)
+                ?? middleware.Target.ToString(); // Class.Method
+
+            InnerBuilder.UseMiddleware<AnalysisMiddleware>(middlewareName);
```
Comment (lodejard) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: Nice use of middleware constructor parameters!

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class AnalysisServiceCollectionExtensions
+    {
+        public static IServiceCollection AddMiddlewareAnalysis(this IServiceCollection services)
+        {
+            // This should prevent AnalysisStartupFilter from being registered more than once.
+            services.TryAddEnumerable(ServiceDescriptor.Singleton<IStartupFilter, AnalysisStartupFilter>());
```
Comment (lodejard) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisServiceCollectionExtensions.cs]: Minor suggestion - might be better as transient - othewise the `AnalysisStartupFilter` instance will be added to the application scope service dictionary for the lifetime of the web app

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisMiddleware.cs]: That's not how we've been doing it anywhere else. E.g. https://github.com/aspnet/Hosting/pull/463/files#diff-aa52cd5a6af48a03a569c8171c3e02abR74

**Comment by Tratcher** [comment]
Updated.

**Comment by Tratcher** [comment]
I changed the default name from `Func<RequestDelegate, RequestDelegate>.Target` to the resulting `RequestDelegate.Target`. This bypassed most of the indirection issues. The only common broken scenario is the `Use(Func<HttpContext, Func<Task>, Task>)` extension. 
https://github.com/aspnet/HttpAbstractions/commit/dc9529c869a44cdf147bfbfccd7a8c19165bd035#diff-54d6cacc4eeb812168e08000576e7b51R21.  I kept the naming convention to deal with issues like this, but removed all of the sugar for it.

**Comment by Tratcher** [comment]
*poke* @avanderhoorn @davidfowl @muratg

**Comment by davidfowl** [comment]
Why do we need to keep the naming convention?

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        public IFeatureCollection ServerFeatures
+        {
+            get { return InnerBuilder.ServerFeatures;}
+        }
+
+        public RequestDelegate Build()
+        {
+            // Add one maker at the end before the default 404 middleware (or any fancy Join middleware).
+            return InnerBuilder.UseMiddleware<AnalysisMiddleware>("EndOfPipeline")
```
Comment (davidfowl) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: If you call Build twice won't it add the AnalysisMiddleware again?

Comment (Tratcher) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: True. We don't ever call build twice though, and worst case you end up with some duplicate events.

Comment (Tratcher) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: This Build is also only ever called on branches that have first called New (e.g. Map). See https://github.com/aspnet/Diagnostics/pull/207/files#diff-9ac90184c9cfc4a15d6622dbce5b5cf6R19

Comment (davidfowl) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: We don't but the API is busted here.

Comment (Tratcher) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: Would you prefer an InvalidOperationException if it's called twice? That seems worse than a duplicate EndOfPipeline event.

I don't think we can do anything else but change from wrapping the inner builder to completely replacing it.

Comment (davidfowl) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: Ok.

**Comment by davidfowl** [comment]
:shipit:

File: src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs]: File a bug

**Comment by davidfowl** [comment]
Why is the CI failing?

**Pull request merged**